src/metrics.jl

```diff
@@ -180,20 +215,20 @@ end
             @simd for I in eachindex(a, b)
                 ai = a[I]
                 bi = b[I]
-                s = eval_reduce(d, s, eval_op(d, ai, bi))
+                s = eval_reduce(d, s, p === nothing ? eval_op(d, ai, bi) : eval_op(d, ai, bi, p[I]))
             end
         else
             for (Ia, Ib) in zip(eachindex(a), eachindex(b))
                 ai = a[Ia]
                 bi = b[Ib]
-                s = eval_reduce(d, s, eval_op(d, ai, bi))
+                s = eval_reduce(d, s, p === nothing ? eval_op(d, ai, bi) : eval_op(d, ai, bi, p[Ia]))
```

Comment:
For correctness I think you'll have to write a separate loop with `(Ia, Ib, Ip) in zip(eachindex(a), eachindex(b), eachindex(p))`. Indeed, let's imagine `a` is 3×2 and `b` is 2×3 then `p` may be 2×3 and if `Base.IndexStyle(a) == CartesianIndexing()` we'll try to call `p[3, 2]` which won't work (I know, that's a contrived example, but...).